_episodes/08-connected-components.md

```diff
@@ -261,156 +319,388 @@ max: 11
+> > of objects. In that case, Numpy array functions turn out to be
+> > very useful because they are much faster. In this example, we can
+> > also use the `np.count_nonzero` function that we have seen earlier
+> > together with the `>` operator to count the objects whose area is
+> > above `min_area`.
+> > ~~~
+> > n = np.count_nonzero(object_areas > min_area)
+> > print("Found", n, "objects!")
+> > ~~~
+> > {: .language-python}
+> >
+> > For all three alternatives, the output is the same and gives the
+> > expected count of 7 objects.
+> >
+> > Functions from Python packages such as Numpy are often more
+> > efficient and require less code to write. It is a good idea to
+> > browse the reference pages of `numpy` and `skimage` to look for an
+> > availabe function that can solve a given task.
 > {: .solution}
+{: .challenge}
+
+> ## Remove small objects (20 min)
+>
+> We might also want to exclude (mask) the small objects when plotting
+>  the labeled image.
 >
-> Now make this change visual:
-> Modify the label image such that objects below a certain area are set to the background label (0).
+> 2. Enhance the `connected_components` function such that it
+> automatically removes objects that are below a certain area that is
+> passed to the function as an optional parameter.
 >
-> > # Solution
+> > ## Solution
+> >
+> >
+> > To remove the small objects from the labeled image, we change the
+> > value of all pixels that belong to the small objects to the
+> > background label 0. One way to do this is to loop over all objects
+> > and set the pixels that match the label of the object to 0.
 > >
 > > ~~~
-> > # iterate over object_ids and modify the `labeled_image` in-place
 > > for object_id, objf in enumerate(object_features, start=1):
-> >     if objf["area"] < 10000:
-> >         labeled_image[labeled_image == object_id] = 0
+> >     if objf["area"] < min_area:
+> >         labeled_image[labeled_image == objf["label"]] = 0
 > > ~~~
 > > {: .language-python}
-> {: .solution}
->
-> Lastly print out the count for the large objects
->
-> > # Solution
+> >
+> > Here Numpy functions can also be used to eliminate `for` loops and
+> > `if` statements. Like above, we can create an array of the small
+> > object labels with the function `np.where`. Then we can use
+> > another Numpy function, `np.isin`, to set the pixels of all small
+> > objects to 0. `np.isin` returns the indices of an array that are
+> > found in the second array that is passed to the function, here
+> > `small_objects`. The loop-free solution then looks like so
 > >
 > > ~~~
-> > # generate a list of objects above a certain area
-> > filtered_list = []
-> > for objf in object_features:
-> >     if objf["area"] > 10000:
-> >         filtered_list.append(objf["area"])
+> > object_areas = np.array([objf["area"] for objf in object_features])
+> > object_labels = np.array([objf["label"] for objf in object_features])
+> > small_objects = object_labels[np.where(object_areas < min_area)]
```

Comment:
```suggestion
> > small_objects = object_labels[object_areas < min_area]
```
Same here - I'm not sure you need the ```np.where```.

Reply:
Remove `np.where` in b878897f2f62261f6c4899459b1ee01dc67b32f4.